File: lib-esphome-client/src/test/java/io/github/arhor/esphome/client/async/noise/NoiseCipherStateTest.java
package io.github.arhor.esphome.client.async.noise;

import io.github.arhor.esphome.client.async.EspHomeProtocolException;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;

class NoiseCipherStateTest {

    @Test
    void encryptsAndDecryptsWithMatchingCipherStates() {
        final var key = new byte[32];
        for (var index = 0; index < 32; index++) {
            key[index] = (byte) (index + 1);
        }
        final var sender = new NoiseCipherState();
        sender.initializeKey(key);
        final var receiver = new NoiseCipherState();
        receiver.initializeKey(key);
        final var associatedData = "ad".getBytes(StandardCharsets.UTF_8);
        final var plaintext = "hello encrypted esphome".getBytes(StandardCharsets.UTF_8);

        final var ciphertext = sender.encryptWithAd(associatedData, plaintext);
        final var decrypted = receiver.decryptWithAd(associatedData, ciphertext);

        assertFalse(java.util.Arrays.equals(ciphertext, plaintext));
        assertArrayEquals(plaintext, decrypted);
    }

    @Test
    void rejectsTamperedCiphertext() {
        final var key = new byte[32];
        for (var index = 0; index < 32; index++) {
            key[index] = (byte) (index + 1);
        }
        final var sender = new NoiseCipherState();
        sender.initializeKey(key);
        final var receiver = new NoiseCipherState();
        receiver.initializeKey(key);
        final var ciphertext = sender.encryptWithAd(NoiseConstants.EMPTY, "hello".getBytes(StandardCharsets.UTF_8));
        ciphertext[ciphertext.length - 1] = (byte) (ciphertext[ciphertext.length - 1] ^ 0x01);

        assertThrows(EspHomeProtocolException.class, () -> receiver.decryptWithAd(NoiseConstants.EMPTY, ciphertext));
    }

    @Test
    void usesNonceCounterForSubsequentMessages() {
        final var key = new byte[32];
        for (var index = 0; index < 32; index++) {
            key[index] = (byte) (index + 1);
        }
        final var cipher = new NoiseCipherState();
        cipher.initializeKey(key);

        final var first = cipher.encryptWithAd(NoiseConstants.EMPTY, "same".getBytes(StandardCharsets.UTF_8));
        final var second = cipher.encryptWithAd(NoiseConstants.EMPTY, "same".getBytes(StandardCharsets.UTF_8));

        assertFalse(java.util.Arrays.equals(first, second));
    }
}

File: services/lib-esphome-client/src/test/kotlin/io/github/arhor/esphome/client/internal/noise/NoiseCipherStateTest.kt
package io.github.arhor.esphome.client.internal.noise

import io.github.arhor.esphome.client.EspHomeProtocolException
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertFalse
import kotlin.test.assertFailsWith

class NoiseCipherStateTest {

    @Test
    fun `encrypts and decrypts with matching cipher states`() {
        val key = ByteArray(32) { index -> (index + 1).toByte() }
        val sender = NoiseCipherState().apply { initializeKey(key) }
        val receiver = NoiseCipherState().apply { initializeKey(key) }
        val associatedData = "ad".encodeToByteArray()
        val plaintext = "hello encrypted esphome".encodeToByteArray()

        val ciphertext = sender.encryptWithAd(associatedData, plaintext)
        val decrypted = receiver.decryptWithAd(associatedData, ciphertext)

        assertFalse(ciphertext.contentEquals(plaintext))
        assertContentEquals(plaintext, decrypted)
    }

    @Test
    fun `rejects tampered ciphertext`() {
        val key = ByteArray(32) { index -> (index + 1).toByte() }
        val sender = NoiseCipherState().apply { initializeKey(key) }
        val receiver = NoiseCipherState().apply { initializeKey(key) }
        val ciphertext = sender.encryptWithAd(ByteArray(0), "hello".encodeToByteArray())
        ciphertext[ciphertext.lastIndex] = (ciphertext.last().toInt() xor 0x01).toByte()

        assertFailsWith<EspHomeProtocolException> {
            receiver.decryptWithAd(ByteArray(0), ciphertext)
        }
    }

    @Test
    fun `uses nonce counter for subsequent messages`() {
        val key = ByteArray(32) { index -> (index + 1).toByte() }
        val cipher = NoiseCipherState().apply { initializeKey(key) }

        val first = cipher.encryptWithAd(ByteArray(0), "same".encodeToByteArray())
        val second = cipher.encryptWithAd(ByteArray(0), "same".encodeToByteArray())

        assertFalse(first.contentEquals(second))
    }
}
